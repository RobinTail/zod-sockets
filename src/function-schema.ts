import type { $InferInnerFunctionType } from "zod/v4/core";
import { z } from "zod/v4";

/** @link https://github.com/colinhacks/zod/issues/4143#issuecomment-2931729793 */
export const functionSchema = <IN extends z.ZodTuple, OUT extends z.ZodType>(
  input: IN,
  output: OUT,
  params?: { path?: PropertyKey[] },
): z.ZodType<(...args: z.output<IN>) => z.output<OUT>> => {
  const schema = z.function({ input, output });
  return z.custom().transform((arg, ctx) => {
    if (typeof arg !== "function") {
      ctx.addIssue({ ...params, message: "Expected function" });
      return z.NEVER;
    }
    return schema.implement(arg as $InferInnerFunctionType<IN, OUT>);
  });
};
