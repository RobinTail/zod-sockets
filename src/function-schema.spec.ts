import { z } from "zod/v4";
import { functionSchema } from "./function-schema";

describe("Function schema", () => {
  const schema = functionSchema(z.tuple([z.string(), z.number()]), z.boolean());

  test("should be instance of ZodPipe", () => {
    expect(schema).toBeInstanceOf(z.ZodPipe);
  });

  test("should accept valid function", () => {
    const impl = (a: string, b: number) => Boolean(a || b);
    const parsed = schema.parse(impl);
    expect(typeof parsed).toBe("function");
    expectTypeOf(parsed).toEqualTypeOf<(a: string, b: number) => boolean>();
    expect(parsed("123", 456)).toBe(true);
  });

  test("should not accept less arguments", () => {
    const impl = (a: string) => Boolean(a);
    const parsed = schema.parse(impl) as (a: string) => boolean;
    expect(() => parsed("123")).toThrowErrorMatchingSnapshot();
  });

  test("should not accept no arguments", () => {
    const impl = () => false;
    const parsed = schema.parse(impl) as () => boolean;
    expect(() => parsed()).toThrowErrorMatchingSnapshot();
  });

  test("should not accept wrong output", () => {
    const impl = (a: string, b: number) => `${a}${b}`;
    const parsed = schema.parse(impl);
    expect(() => parsed("123", 456)).toThrowErrorMatchingSnapshot();
  });

  test("should not accept wrong args", () => {
    const impl = (a: string, b: number) => Boolean(a || b);
    const parsed = schema.parse(impl);
    expect(() =>
      parsed(true as unknown as string, 456 as unknown as number),
    ).toThrowErrorMatchingSnapshot();
  });

  test("should not accept not-a-function", () => {
    expect(() => schema.parse("test")).toThrowErrorMatchingSnapshot();
  });
});
