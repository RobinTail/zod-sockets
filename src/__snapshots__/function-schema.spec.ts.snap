// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Function schema > should not accept less arguments 1`] = `
$ZodError({
  "message": "[
  {
    "expected": "number",
    "code": "invalid_type",
    "path": [
      1
    ],
    "message": "Invalid input: expected number, received undefined"
  }
]",
})
`;

exports[`Function schema > should not accept no arguments 1`] = `
$ZodError({
  "message": "[
  {
    "origin": "array",
    "code": "too_small",
    "minimum": 2,
    "path": [],
    "message": "Too small: expected array to have >2 items"
  }
]",
})
`;

exports[`Function schema > should not accept not-a-function 1`] = `
ZodError({
  "issues": [
    {
      "code": "custom",
      "message": "Expected function",
      "path": [],
    },
  ],
})
`;

exports[`Function schema > should not accept wrong args 1`] = `
$ZodError({
  "message": "[
  {
    "expected": "string",
    "code": "invalid_type",
    "path": [
      0
    ],
    "message": "Invalid input: expected string, received boolean"
  }
]",
})
`;

exports[`Function schema > should not accept wrong output 1`] = `
$ZodError({
  "message": "[
  {
    "expected": "boolean",
    "code": "invalid_type",
    "path": [],
    "message": "Invalid input: expected boolean, received string"
  }
]",
})
`;
